# CMSIP

Detecting differential 5hmC regions from CMS-IP sequencing data.

Source URL: [https://github.com/lijinbio/cmsip](https://github.com/lijinbio/cmsip)

![Workflow of CMSIP.](cmsip_flowchart.png)

## Installation

### Dependencies

- bsmap

`bsmap` is a component in the MOABS package. See more at MOABS ([https://github.com/sunnyisgalaxy/moabs](https://github.com/sunnyisgalaxy/moabs)).

- samtools: [http://samtools.sourceforge.net](http://samtools.sourceforge.net)

- bedtools: [https://bedtools.readthedocs.io](https://bedtools.readthedocs.io)

- kentUtils: [https://github.com/ENCODE-DCC/kentUtils](https://github.com/ENCODE-DCC/kentUtils)

## Example configuration file and description

```
sampleinfo:
  - sampleid: K1
    group: K
    filenames:
      - DKO1.CMS.fq.gz
  - sampleid: K2
    group: K
    filenames:
      - DKO2.CMS.fq.gz
  - sampleid: W1
    group: W
    filenames:
      - WT1.CMS.fq.gz
  - sampleid: W2
    group: W
    filenames:
      - WT2.CMS.fq.gz
  - sampleid: IK1
    group: IK
    filenames:
      - DKO.input.fastq.gz
  - sampleid: IW1
    group: IW
    filenames:
      - WT.input.fastq.gz
groupinfo:
  group1: K
  group2: W
resultdir: /data/result
aligninfo:
  reference: mm10.fa
  usespikein: True
  spikein: hg38.fa
  fastqdir: /data/fastq
  statfile: qcstats.txt
  barplotinfo:
    outfile: qcstats_twsn_barplot.pdf
    height: 5
    width: 5
  numthreads: 24
  verbose: True
genomescaninfo:
  readextension: True
  fragsize: 100
  windowfile: /data/result/mm10_w200.bed
  referencename: mm10
  windowsize: 200
  readscount: True
  counttablefile: /data/result/extTrue_cntTrue_w200.txt.gz
  verbose: True
dhmrinfo:
  method: gtest
  meandepth: 1
  testfile: /data/result/gtest_extTrue_cntTrue_w200.txt.gz
  qthr: 0.05
  maxdistance: 0
  dhmrfile: /data/result/gtest_extTrue_cntTrue_w200.dhmr.gz
  numthreads: 24
  nsplit: 1000
  verbose: True
  keepNA: True
useinput: True
inputinfo:
  group1: IK
  group2: IW
  method: gtest
  qthr: 0.05
  testfile1: /data/result/gtest_extTrue_cntTrue_w200_G1VsInput.txt.gz
  dhmrfile1: /data/result/gtest_extTrue_cntTrue_w200_G1VsInput.dhmr.gz
  testfile2: /data/result/gtest_extTrue_cntTrue_w200_G2VsInput.txt.gz
  dhmrfile2: /data/result/gtest_extTrue_cntTrue_w200_G2VsInput.dhmr.gz
  inputfilterfile: /data/result/extTrue_cntTrue_w200_inputfilter.txt.gz
  verbose: True
```

### `sampleinfo`

This block stores detailed metadata information of samples.

### `groupinfo`

This block lists the interested comparison.  The alternative hypothesis is true difference in means of `group1` and `group2` is less than 0.

### `aligninfo`

Options and data information required for alignment.

- reference

The FASTA file for the reference genome, such as hg38.fa.gz.

- spikein

The FASTA file for the spike-in genome, such as mm10.fa.gz.

- windowfile: hg38_w100.bed

The genome in window bins. This window bin file can be generated by using bedtools. E.g.

```
bedtools makewindows -g <(fetchChromSizes hg38) -w 100 > hg38_w100.bed
```

- windowsize: 100

Window size for creating bins.

- fastqdir: test_data

Root directory with raw FASTQ files.

-  outdir

Root output directory for temporary and final result files.

- statfile

QC statistics file. Default is at outdir/qcstats.txt. If this file exists, QC step will be skipped, and size factors will be parsed for the existing QC statistical file. Otherwise, QC step will run to generate the statistics file.

- cnttablefile

Region count table file. Default is at outdir/meancovtable.txt.gz. If this file exists, counting step will be skipped, and the existing count table file will be used for downstream statistical testing. Otherwise, counting step will execute to generate the count table file.

- ttestfile

The statistical testing result file. Default is at outdir/t.test.txt. If this file exists, no more task will run. Otherwise, statistical testing will run on the count table using t-test.


